src/persist/src/indexed/mod.rs

```diff
@@ -689,11 +714,93 @@ impl<L: Log, B: Blob> Indexed<L, B> {
             self.blob.delete_unsealed_batch(&batch)?;
         }
 
-        for batch in deleted_trace_batches {
-            self.blob.delete_trace_batch(&batch)?;
+        Ok(())
+    }
+
+    fn compact_trace_maintenance_reqs(&mut self) -> Result<Vec<MaintenanceReq>, Error> {
+        let mut reqs = vec![];
+        for (id, arrangement) in self.state.arrangements.iter() {
+            // Skip over the arrangement if it already has a compaction request in flight.
+            //
+            // TODO: note that this works because right now the connection between Indexed
+            // and Maintainer is infallible, and the two cannot fail independently of each
+            // other. If that changes, we will have to revisit the policy around waiting
+            // arbitrarily long for maintenance responses and most likely institute
+            // a timeout.
+            if self.in_flight_trace_compactions.contains_key(&id) {
+                continue;
+            }
+            if let Some(req) = arrangement.trace_next_compact_req()? {
+                reqs.push(MaintenanceReq::CompactTrace((*id, req.clone())));
+                self.in_flight_trace_compactions.insert(*id, req);
+            }
+        }
+        Ok(reqs)
+    }
+
+    fn do_handle_compact_trace_res(&mut self, id: Id, res: CompactTraceRes) {
```

Comment:
that doesn't work unfortunately because `apply_batched_cmd` is defined on Indexed and not AppliedState. I think `handle_compact_trace_response` on Arrangement is really the only bit of AppliedState specific logic, and that didn't seem worth moving into its own function.